include/swift/Runtime/Mutex.h

```diff
@@ -0,0 +1,313 @@
+  ///
+  /// For example the following mutates value while holding the mutex lock.
+  ///
+  ///   mutex.lock([&value] { value++; });
+  ///
+  /// Precondition: Mutex unlocked by this thread, undefined otherwise.
+  template<typename CriticalSection>
+  void lock(CriticalSection criticalSection) const {
+    ScopedLockImplementation guard(Impl);
+    criticalSection();
+  }
+
+  /// Acquires lock before calling the supplied critical section. If critical
+  /// section returns `true` then it will wait on the supplied condition and
+  /// call critical section again when wait returns (again holding the lock).
+  /// If critical section returns `false` it will no longer wait, it will
+  /// release the lock and return (e.g. lockOrWait returns).
+  ///
+  /// For example the following will loop waiting on condition until
+  /// `value > 0`. It will then "consume" that value and stop looping.
+  /// ...all while being correctly protected by mutex.
+  ///
+  ///   mutex.lockOrWait(condition, [&value] {
+  ///     if (value > 0) {
+  ///       value--;
+  ///       return false;
+  ///     }
+  ///    return true;
+  ///   });
+  ///
+  /// Precondition: Mutex unlocked by this thread, undefined otherwise.
+  template<typename CriticalSection>
+  void lockOrWait(const Condition &condition,
+                  CriticalSection criticalSection) const {
+    ScopedLockImplementation guard(Impl);
+    while ( criticalSection() ) {
+      Impl.wait(condition);
+    }
+  }
+
+  /// Acquires lock before calling the supplied critical section and on return
+  /// from critical section it notifies one waiter of supplied condition and
+  /// then releases the lock.
+  ///
+  /// For example the following mutates value while holding the mutex lock and
+  /// then notifies one condition waiter about this change.
+  ///
+  ///   mutex.lockAndNotifyOne([&value] { value++; });
+  ///
+  /// Precondition: Mutex unlocked by this thread, undefined otherwise.
+  template<typename CriticalSection>
+  void lockAndNotifyOne(const Condition &condition,
+                        CriticalSection criticalSection) const {
+    ScopedLockImplementation guard(Impl);
+    criticalSection();
+    condition.notifyOne();
+  }
+
+  /// Acquires lock before calling the supplied critical section and on return
+  /// from critical section it notifies all waiters of supplied condition and
+  /// then releases the lock.
+  ///
+  /// For example the following mutates value while holding the mutex lock and
+  /// then notifies all condition waiters about this change.
+  ///
+  ///   mutex.lockAndNotifyAll([&value] { value++; });
+  ///
+  /// Precondition: Mutex unlocked by this thread, undefined otherwise.
+  template<typename CriticalSection>
+  void lockAndNotifyAll(const Condition &condition,
+                        CriticalSection criticalSection) const {
+    ScopedLockImplementation guard(Impl);
+    criticalSection();
+    condition.notifyAll();
+  }
+
+protected:
```

Comment:
As currently implement ScopedLock and ScopedUnlock (friends of Mutex) access Impl.